### ISSUE TITLE
DirChangeNotify misses changes

### ISSUE THREAD

**Issue by B4dM4n** [rclone/rclone#2006]
As I read through #1937 I became aware of the following statement about `DirChangeNotify` in the drive remote from @remusb:
> moving a file doesn't seem to expire both parent folders, only the destination. At the moment this makes things a bit tricky as cache would best work if it doesn't expire its entries at all but it seems some cases aren't included in the notifications

I tried moving some files between folders using a master build of rclone to replicate the behavior.
My setup was the following:
```sh
$ rclone mount -vv --dir-cache-time 1h mydrive: /mnt/a
$ rclone mount -vv --dir-cache-time 1h mydrive: /mnt/b
$ find /mnt/a
/mnt/a/folder1
/mnt/a/folder1/file1
/mnt/a/folder1/file2
/mnt/a/folder2
$ find /mnt/b
/mnt/b/folder1
/mnt/b/folder1/file1
/mnt/b/folder1/file2
/mnt/b/folder2
$ mv /mnt/a/folder1/file1 /mnt/a/folder2
# wait for "Checking for changes on remote" in /mnt/b
$ find /mnt/b
/mnt/b/folder1
/mnt/b/folder1/file1
/mnt/b/folder1/file2
/mnt/b/folder2
/mnt/b/folder2/file1
```

> What is your rclone version (eg output from `rclone -V`)

rclone v1.39-072-g33b19c13
- os/arch: linux/amd64
- go version: go1.9.2

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux (NixOS)
> Which cloud storage system are you using? (eg Google Drive)

Google Drive

While digging through `backend/drive/drive.go` i found the source of this bug. In [drive.go#1153](https://github.com/ncw/rclone/blob/master/backend/drive/drive.go#L1153) `DirChangeNotify` tries to use `f.dirCache.GetInv` to get the old path for the changed file. This can never work, because `dirCache` is, as the name implies, only a directory cache, not a file cache.

~~The amazondrive remote is using the same method in  [amazonclouddrive.go#1264](https://github.com/ncw/rclone/blob/master/backend/amazonclouddrive/amazonclouddrive.go#L1264). It should be affected as well.~~ amazondrive is not affected

What would be the best fix for this bug? Should `dircache.DirCache` be changed into a full file tree cache?

**Comment by ncw** [comment]
Excellent work to find the cause of that bug!

> What would be the best fix for this bug? Should dircache.DirCache be changed into a full file tree cache?

That would be rather expensive in terms of memory I think.

(Aside: I don't really like the DirCache - at some point I'd like to upgrade rclone's concept of directories from strings into an object so it holds an ID and then I think we could get rid of it.)

What I'd like to do is use the changes to find the IDs of the parent directories the old and the new, before and after the move. Looking [in the docs](https://developers.google.com/drive/v2/reference/changes#resource) there doesn't appear to be a way of geting a list of the old parents of a file though.

Does a deleted from the old parent event get fired?

@breunigs any ideas - I think you wrote this bit of code originally?

**Comment by B4dM4n** [comment]
Unfortunately there is no event for the old or the new directory in Google Drive.

I checked the behavior for Amazon Drive and the bug does not appear there. The ACD API emits events for folder1, folder2 and file1 in the above example.

**Comment by ncw** [comment]
> Unfortunately there is no event for the old or the new directory in Google Drive.

I tried this about 10 times, and once I got two events

```
2018/01/24 14:19:24 DEBUG : folder1: forgetting directory cache
2018/01/24 14:19:24 DEBUG : folder2: forgetting directory cache
```

I haven't been able to reproduce that so not 100% sure I didn't fat finger something.

**Comment by ncw** [comment]
It occurs to me that we do in fact have a cache of exactly the objects with their IDs that we need to clear - in the VFS directory cache.

We'd need to change the DirChangeNotify interface somewhat... maybe it should return paths and object  IDs which need to be cleared.

We'd need also to add an ID method to Object (but I've been planning that anyway).

What do you think?

**Comment by remusb** [comment]
I was thinking of doing something along these lines for this issue: https://github.com/ncw/rclone/issues/1937

```

**Comment by remusb** [comment]
Let me know if I can start working on something like this if you guys haven't done it yet. I don't want us to duplicate work ;)

**Comment by ncw** [comment]
Does anyone know if this issue is still valid?

**Comment by B4dM4n** [comment]
Yes, this issue is still present.

Moving a file `file.txt` from folder `a` to `b` will only trigger a notification for `file.txt` with it's parent set to `b`.

**Comment by ivandeex** [comment]
@B4dM4n  This problem is reported against obsolete rclone version. Can you reproduce it with rclone 1.54?

**Comment by ivandeex** [comment]
This was fixed by #2606
